File: app/javascript/channels/chatroom_channel.js
import consumer from "channels/consumer"

const appChat = consumer.subscriptions.create("ChatroomChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    const messages = document.getElementById('messages')
    messages.insertAdjacentHTML('afterbegin', data['message'])
  },

  talk(data) {
    return this.perform('talk', data)
  }
});

document.getElementById('content').addEventListener('keyup', (e) => {	(4)
  if (e.key === 'Enter') {
    const sender = document.getElementById('sender')
    appChat.talk({sender: sender.value, content: e.target.value});
    e.target.value = ''
    e.preventDefault()
  }
})

File: app/channels/chatroom_channel.rb
class ChatroomChannel < ApplicationCable::Channel
  def subscribed
    # chatroom_channelからのストリーミングを開始する
    stream_from "chatroom_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  # クライアントがメッセージ送信に使うメソッド
  def talk(data)
    # クライアントからのデータをもとにmessageモデルを組み立てる（現在日時を使用）
    message = Message.new
    message.published = Time.now
    message.sender = data['sender']
    message.content = data['content']
    Message.create!(published: message.published, sender: message.sender, content: message.content)
  end
end
